File: goormlevel/167345/1. 단풍나무/단풍나무.py
import sys

input = sys.stdin.readline
write = sys.stdout.write

def solve():
	n = int(input().rstrip())
	nlist = [list(map(int, input().split())) for _ in range(n)]

	dh = [1, 0, 0, -1]
	dw = [0, 1, -1, 0]

	ice = []
	for i in range(n):
		for j in range(n):
			if nlist[i][j] > 0:
				ice.append((i, j))

	answer = 0
	while ice:
		next_ice = []
		melt_info = []

		for i, j in ice:
			count = 0
			for k in range(4):
				row = i + dh[k]
				col = j + dw[k]
				
				if (row >= 0 and row < n) and (col >= 0 and col < n) and (nlist[row][col] == 0):
					count += 1

			if nlist[i][j] - count > 0:
				next_ice.append((i, j))
				melt_info.append((i, j, nlist[i][j] - count))
			else:
				melt_info.append((i, j, 0))

		for i, j, val in melt_info:
			nlist[i][j] = val

		ice = next_ice
		answer += 1

	write(str(answer) + "\n")

if __name__ == "__main__":
	solve()

File: goormlevel/167345/1. 단풍나무/단풍나무.cpp
#include <iostream>
#include <vector>

using namespace std;

struct Melt {
	int x;
	int y;
	int val;
};

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);
	
	int n = 0;
	if (!(cin >> n)) return 0;

	vector<vector<int>> nlist(n, vector<int>(n));
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			if (!(cin >> nlist[i][j])) return 0;
		}
	}

	vector<pair<int, int>> ice;
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			if (nlist[i][j] > 0) {
				ice.push_back({i, j});
			}
		}
	}

	int dh[4] = {1, 0, 0, -1};
	int dw[4] = {0, 1, -1, 0};

	int answer = 0;
	while (!ice.empty()) {
		vector<pair<int, int>> next_ice;
		vector<Melt> melt_info;

		for (const auto& p : ice) {
			int x = p.first;
			int y = p.second;
			
			int count = 0;
			for (int i = 0; i < 4; i++) {
				int row = x + dh[i];
				int col = y + dw[i];

				if ((0 <= row && row < n) && (0 <= col && col < n) && nlist[row][col] == 0) {
					count++;
				}
			}

			if (nlist[x][y] - count > 0) {
				next_ice.push_back({x, y});
				melt_info.push_back({x, y, nlist[x][y] - count});
			}
			else {
				melt_info.push_back({x, y, 0});
			}
		}

		for (const auto& l : melt_info) {
			nlist[l.x][l.y] = l.val;
		}

		ice = next_ice;
		answer++;
	}

	cout << answer << "\n";

	return 0;
}